File: 454.4-sum-ii.cpp
/*
 * @lc app=leetcode id=454 lang=cpp
 *
 * [454] 4Sum II
 */
#include <unordered_map>
#include <vector>
using namespace std;
// @lc code=start
class Solution
{
public:
    int fourSumCount(vector<int> &nums1, vector<int> &nums2, vector<int> &nums3, vector<int> &nums4)
    {
        unordered_map<int, int> count;
        int result = 0;
        int len = nums1.size();
        for (int i = 0; i != len; ++i)
        {
            for (int j = 0; j != len; ++j)
            {
                ++count[nums1[i] + nums2[j]];
            }
        }
        for (int i = 0; i != len; ++i)
        {
            for (int j = 0; j != len; ++j)
            {
                result += count[-nums3[i] - nums4[j]];
            }
        }
        return result;
    }
};
// @lc code=end
// class Solution
// {
// public:
//     int fourSumCount(vector<int> &nums1, vector<int> &nums2, vector<int> &nums3, vector<int> &nums4)
//     {
//         unordered_map<int, int> count_1, count_2;
//         int result = 0;
//         int len = nums1.size();
//         for (int i = 0; i != len; ++i)
//         {
//             for (int j = 0; j != len; ++j)
//             {
//                 ++count_1[nums1[i] + nums2[j]];
//                 ++count_2[nums3[i] + nums4[j]];
//             }
//         }
//         for (auto i = count_1.begin(); i != count_1.end(); ++i)
//         {
//             if (count_2.count(-i->first))
//             {
//                 result += i->second * count_2[-i->first];
//             }
//         }

//         return result;
//     }
// };

File: java/454.4-sum-ii.java
import java.util.HashMap;
import java.util.Map;

/*
 * @lc app=leetcode id=454 lang=java
 *
 * [454] 4Sum II
 */

// @lc code=start
class Solution {
    public int fourSumCount(int[] nums1, int[] nums2, int[] nums3, int[] nums4) {
        Map<Integer, Integer> count = new HashMap<>();
        for (var n1 : nums1) {
            for (var n2 : nums2) {
                count.put(n1 + n2, count.getOrDefault(n1 + n2, 0) + 1);
            }
        }
        int result = 0;
        for (var n3 : nums3) {
            for (var n4 : nums4) {
                result += count.getOrDefault(-(n3 + n4), 0);

            }
        }
        return result;
    }
}
// @lc code=end
